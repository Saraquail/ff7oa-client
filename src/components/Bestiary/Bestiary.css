

.page-title {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.holy-left, .holy-right {
  width: 3em;
  height: 3em;
  margin-left: 1em;
  margin-right: 1em;
}

.holy-right {
  transform: scaleX(-1);
}

.bestiary {
  background-image: linear-gradient(to bottom right,  rgb(14,51,95), rgb(121,172,229), rgb(30,103,187), rgb(14,51,95));
  padding-bottom: 5em;
}