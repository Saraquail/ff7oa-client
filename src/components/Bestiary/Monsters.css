* {
  text-align: center;
}

#expand {
  font-size: 14px;
  width: 15em;
}

#expand:hover, #expand:focus {
  color: #26437a;
}

.monster {
  border: solid 1px;
  border-radius: 25px;
  margin: auto;
  margin-bottom: 2em;
  padding: 1em;
  width: 50%
}

