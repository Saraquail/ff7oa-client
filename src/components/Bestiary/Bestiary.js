import React, { Component } from 'react'
import Nav from '../Nav/Nav'
import OptionsForm from '../OptionsForm/OptionsForm'
import MonsterApiService from '../../services/monster-api-service'
import Monsters from './Monsters'
import holy from '../../images/White_Materia.png'
import './Bestiary.css'

class Bestiary extends Component {
  state = {
    monsters: [],
    sort: '',
    param: ''
  }

  componentDidMount() {
    MonsterApiService.getMonsters()
      .then (data => {
        this.setState({ monsters: data })
      })
      .catch()
  }

  renderMonsters = () => {
    let { sort, param, term } = this.state    
    let allMonsters = this.state.monsters
    if(!sort && !term) {
      allMonsters = this.state.monsters
    }

    if(sort) {
      allMonsters = this.renderSorted(sort)
    }

    if(term) {
      allMonsters = this.renderSearched(param, term)
    }

    if(!allMonsters) {
      return 'No Results Found'
    }

    return allMonsters.map(mon =>
      <Monsters
        id={mon.id}
        key={mon.id}
        name={mon.name}
        hp={mon.hp}
        mp={mon.mp}
        exp={mon.exp}
        gil={mon.gil}
        weakness={mon.weakness}
        strength={mon.strength}
        location={mon.location}
        level={mon.level}
        steal={mon.steal}
        drops={mon.drops}
        enemy_skill={mon.enemy_skill}
        user_name={mon.user_name}
      />
    )
  }

  renderSearched = (param, term) => {
    let allMonsters = this.state.monsters
    let mons = []

    if(!term || !param) {
      return allMonsters
    }

    if(term) {
      param = param.toLowerCase()
      term = term.toLowerCase()

      for(let i=0; i < allMonsters.length; i++) {
        for (let [key, value] of Object.entries(allMonsters[i])) {
          if(key === param && value.toLowerCase().includes(term)) {
            mons.push(allMonsters[i])
            return mons
          }
        }
      }
    }
  }

  renderSorted = (sort) => {
    let allMonsters = this.state.monsters
    let sortby = sort.toLowerCase()

    allMonsters.sort((a, b) => {
        return a[sortby] > b[sortby] ? 1 : a[sortby] < b[sortby] ? -1 : 0
    })
    return allMonsters
  }

  handleSort = ev => {
    let sort = ev.target.value
    this.setState({
      sort: sort
    })
  }

  handleReset = () => {
    this.setState({
      params: '',
      sort: ''
    })
    this.renderMonsters()
    document.getElementById("search-form").reset();

  }

  handleSearch = ev => {
    ev.preventDefault()
    let term = ev.target.search.value
    let param = ev.target['filter-by'].value
    this.setState({
      param: param,
      term: term
    })
  }

  render() {
    return (
      <div>
        <Nav />
        <OptionsForm handleReset={this.handleReset} handleSort={this.handleSort} handleSearch={this.handleSearch} />
        <div className="page-title">
          <img src={holy} alt="a pale green orb of holy materia from final fantasy 7" className="materia-img holy-left" />
          <h1>Bestiary</h1>
          <img src={holy} alt="a pale green orb of holy materia from final fantasy 7" className="materia-img holy-right" />
        </div>
          {this.renderMonsters()}
      </div>
    )
  }
}

export default Bestiary