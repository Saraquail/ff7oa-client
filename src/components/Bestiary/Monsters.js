import React, { Component } from 'react'
import { HashLink as Link } from 'react-router-hash-link';
import SaveGuideForm from '../SaveGuideForm/SaveGuideForm'
import './Monsters.css'

class Monsters extends Component {
  state = {
    showModal: false,
    selectedid: '',
    expanded: false,
    sort: ''
  }

  handleExpand = () => {
    this.setState(prevState => ({
      expanded: !prevState.expanded
    }))
  }

  handleOpenModal = (e) => {
    let id = e.target.value
    this.setState({
      showModal: true,
      selectedid: id
    })
  }

  handleCloseModal = () => {
    this.setState({
      showModal: false
    })
  }

  renderAllMons = () => {

    const mon = this.props
    let id = this.state.selectedid
    if(!mon) {
      return <p>Loading Monsters</p>
    }

    return(
      <section className="monsterList" >
        <div className="monster">
          <div className="collapsed">    
            <h2 id={mon.id}>{mon.name}</h2>
            <button value={mon.id}
              onClick={e => this.handleOpenModal(e)}
              > Add to My PHS</button>
            <Link to={'#' + mon.id} id="expand">
              <button id="expand" onClick={this.handleExpand}>
              Click for more/less info
              </button>
            </Link> 
            <p> Level: {mon.level} </p>
            <p> Added By: {mon.user_name} </p>
            <p> HP: {mon.hp} MP: {mon.mp} </p>
            <p> Exp: {mon.exp} Gil: {mon.gil} </p>
            <p> Strength: {mon.strength}</p>
            <p> Weakness: {mon.weakness}</p>
          </div>
          {this.state.expanded 
            ? <div className="expanded">
              <p> Location: {mon.location} </p>
              <p> Steal: {mon.steal} </p>
              <p> Drop: {mon.drops} </p>
              <p> Enemy_skill: {mon.enemy_skill} </p>
            </div>
            : ''}
        </div>
      {this.state.showModal ?
        <SaveGuideForm className="modal overlay"  selectedid={id} handleCloseModal={this.handleCloseModal}   onRequestClose={this.handleCloseModal}
        />
      : ''}
      </section>

    )
  }


  render () {

    return (
      <div>
      
      {this.renderAllMons()}
      </div>
    )
    

  }
}

export default Monsters