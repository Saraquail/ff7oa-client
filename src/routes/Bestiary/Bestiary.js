import React, { Component } from 'react';
import OptionsForm from '../../components/OptionsForm/OptionsForm';
import MonsterApiService from '../../services/monster-api-service';
import Monsters from '../../components/Monsters/Monsters';
import holy from '../../images/White_Materia.png';
import { sortBy } from '../../utils';
import './Bestiary.css';

class Bestiary extends Component {
  state = {
    monsters: [],
    sort: '',
    param: '',
    message: '',
  }

  componentDidMount() {
    MonsterApiService.getMonsters()
      .then((monsters) => {
        this.setState({ monsters });
      })
      .catch();
  }

  renderMonsters = () => {
    const {
      sort, param, term, monsters,
    } = this.state;
    let allMonsters = monsters;
    if (!sort && !term) {
      allMonsters = monsters;
    }

    if (sort) {
      allMonsters = this.renderSorted(sort);
    }

    if (term) {
      allMonsters = this.renderSearched(param, term);
    }

    if (!allMonsters) {
      return 'No Results Found';
    }

    return allMonsters.map((mon) => (
      <Monsters
        id={mon.id}
        key={mon.id}
        name={mon.name}
        hp={mon.hp}
        mp={mon.mp}
        exp={mon.exp}
        ap={mon.ap}
        gil={mon.gil}
        weakness={mon.weakness}
        strength={mon.strength}
        location={mon.location}
        level={mon.level}
        steal={mon.steal}
        drops={mon.drops}
        morph={mon.morph}
        enemy_skill={mon.enemy_skill}
      />
    ));
  }

  renderSearched = (param, term) => {
    /* When searching,  give the user feedback that their query is being executed on the backend
especially  for cases in which no results were returned. I
t's not clear if that message is left from the previous search,
or if the current search is also not returning results.
*/
    const { monsters: allMonsters } = this.state;


    // if the user hits search without filling either field, returns all results
    if (!term || !param) {
      return allMonsters;
    }

    if (term) {
      const contains = (monster) => monster[param.toLowerCase()]
        .toLowerCase()
        .indexOf(term.toLowerCase()) !== -1;
      return allMonsters.filter(contains);
    }
    return null;
  }

  renderSorted = (sort) => {
    const { monsters: allMonsters } = this.state;
    const sortby = sort.toLowerCase();

    // if the sort parameter is a string value, sorts alphabetically
    if (sortby === 'location' || sortby === 'name') {
      allMonsters.sort(sortBy(sortby));
    }
    // if the sort parameter is a number value, sorts by
    // number in descending order (the above returns ascending order)
    else if (sortby === 'gil' || sortby === 'exp') {
      allMonsters.sort(sortBy(sortby, true));
    }

    return allMonsters;
  }

  handleSort = (ev) => {
    const sort = ev.target.value;
    this.setState({
      sort,
    });
  }

  handleReset = () => {
    this.setState({
      param: '',
      sort: '',
    });
    this.renderMonsters();
    document.getElementById('search-form').reset();
  }

  handleSearch = (ev) => {
    ev.preventDefault();
    const term = ev.target.search.value;
    const param = ev.target['filter-by'].value;

    this.setState({
      param,
      term,
    });
  }

  render() {
    const { message } = this.state;
    return (
      <div>
        <section className="bestiary">
          <OptionsForm
            requiredMessage={message}
            handleReset={this.handleReset}
            handleSort={this.handleSort}
            handleSearch={this.handleSearch}
          />
          <div className="page-title">
            <img src={holy} alt="a pale green orb of holy materia from final fantasy 7" className="holy-left" />
            <h1>Bestiary</h1>
            <img src={holy} alt="a pale green orb of holy materia from final fantasy 7" className="holy-right" />
          </div>
          {this.renderMonsters()}
        </section>
      </div>
    );
  }
}

export default Bestiary;
