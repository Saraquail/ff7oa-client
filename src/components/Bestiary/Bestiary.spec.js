import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer';

import Bestiary from './Bestiary.js'

describe('<Bestiary/>', () => {

  it('renders without crashing', () => {
  const div = document.createElement('div')

  ReactDOM.render(
  <BrowserRouter>
    <Bestiary />
  </BrowserRouter>, div)

  ReactDOM.unmountComponentAtNode(div)
  })

  it('renders the UI as expected', () => {
    const tree = renderer
  .create(
    <BrowserRouter>
      <Bestiary/>
    </BrowserRouter>
  )
  .toJSON();
    expect(tree).toMatchSnapshot();  
  })
})
