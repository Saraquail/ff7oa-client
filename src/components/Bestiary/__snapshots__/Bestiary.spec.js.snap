// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bestiary/> renders the UI as expected 1`] = `
<div>
  <nav>
    <div
      className="header-image-container"
    >
      <img
        alt="green orb of magic materia from final fantasy 7"
        className="materia-img"
        src="Magic_Materia.png"
      />
      <img
        alt="purple orb of complete materia from final fantasy 7"
        className="materia-img"
        src="Complete_Materia.png"
      />
      <img
        alt="yellow orb of command materia from final fantasy 7"
        className="materia-img"
        src="Command_Materia.png"
      />
      <img
        alt="blue orb of support materia from final fantasy 7"
        className="materia-img"
        src="Support_Materia.png"
      />
      <img
        alt="red orb of summoning materia from final fantasy 7"
        className="materia-img"
        src="Summoning_Materia.png"
      />
    </div>
    <ul>
      <li>
        <a
          href="/bestiary"
          onClick={[Function]}
        >
          Bestiary
        </a>
      </li>
      <li>
        <a
          href="/PHS"
          onClick={[Function]}
        >
          My PHS
        </a>
      </li>
      <li>
        <a
          href="/"
          onClick={[Function]}
        >
          Login
        </a>
      </li>
    </ul>
  </nav>
  <section
    className="bestiary"
  >
    <section
      id="bestiary-options"
    >
      <form
        id="search-form"
        name="search-form"
        onSubmit={[Function]}
      >
        <label
          htmlFor="filter-by"
        >
          Search by:
        </label>
        <select
          defaultValue=""
          id="filter-by"
          name="filter-by"
        >
          <option>
            -select-
          </option>
          <option>
            Location
          </option>
          <option>
            Name
          </option>
          <option>
            Drops
          </option>
          <option>
            Steal
          </option>
          <option>
            Enemy Skill
          </option>
        </select>
        <label
          htmlFor="search"
        >
          Search:
        </label>
        <input
          id="search"
          name="search"
          type="text"
        />
        <button
          type="submit"
        >
          Search
        </button>
        <button
          onClick={[Function]}
        >
          Reset
        </button>
      </form>
      <div
        className="sort-and-add"
      >
        <label
          htmlFor="sort-by"
        >
          Sort by:
        </label>
        <select
          defaultValue=""
          id="sort-by"
          name="sort-by"
          onChange={[Function]}
        >
          <option>
            -select-
          </option>
          <option>
            Name
          </option>
          <option>
            GIL
          </option>
          <option>
            Location
          </option>
          <option>
            EXP
          </option>
        </select>
        <a
          href="/monster-form"
          id="add-monster-button"
          onClick={[Function]}
        >
          <button
            id="add-monster-button"
          >
            Add New
          </button>
        </a>
      </div>
    </section>
    <div
      className="page-title"
    >
      <img
        alt="a pale green orb of holy materia from final fantasy 7"
        className="holy-left"
        src="White_Materia.png"
      />
      <h1>
        Bestiary
      </h1>
      <img
        alt="a pale green orb of holy materia from final fantasy 7"
        className="holy-right"
        src="White_Materia.png"
      />
    </div>
  </section>
</div>
`;
